src/fhirtypes/ElementDefinition.ts

```diff
@@ -467,21 +477,36 @@ export class ElementDefinition {
       this.definition = rule.definition;
     }
 
+    // Special handling is required for the 'base' and 'constraint' attributes for newly
+    // added elements.
+    // The 'base' attribute is required for the snapshot and is optional in the differential.
+    // The 'constraint' attribute is optional for both snapshot and differential.
+    // REF: http://hl7.org/fhir/elementdefinition.html#interpretation
+    // However, convention, based on looking at FHIR resources, is to exclude the 'base'
+    // attribute from the differential and to only include constraint attributes when
+    // specifically defined for an element.
+    // The 'elements' collection in the FSH StructureDefinition is split out into snapshot
+    // and differential in the StructureDefinition.toJSON() method. Therefore, to achieve the
+    // observed convention, we place the 'base' and the "base" element's constraints into
+    // temp holding properties. During execution of the StructureDefinition.toJSON() method,
+    // we use these temp holding properties to properly set the 'base' and 'constraint'
+    // attributes in the snapshot elements.
```

Comment:
I think we might be able to accomplish this without adding new temporary properties.

Since this function (`applyAddElementRule`) is always called on a brand new element (w/ no changes yet applied), I think we can safely set `base` and `constraint` at the very top of this function and then call `this.captureOriginal()` _before_ setting any of the other things.  This will have the affect of treating them as part of the original definition -- which will keep them in the snapshot but suppress them from the differential.

We already treat `constraint` as an "additive" property ([implementation](https://github.com/joepaquette/sushi/blob/5bb3cd8d0b5c9bdef90ba641abfc298d54dc19fe/src/fhirtypes/ElementDefinition.ts#L389-L396)), which means that when we generate the differential, we only include those constraints that are new (i.e., not in the "original").  So following this approach we should also still get the desired behavior of the differential only including the constraints that were directly set on the element.

Assuming this approach works (and I haven't missed something important), I think it would be a good bit cleaner (and more efficient).

Reply:
It did work and the code is definitely much cleaner. Thanks for the help!